databricks/azure_auth.go

```diff
@@ -93,93 +102,88 @@ func (a *AzureAuth) getWorkspaceID(config *service.DBApiClientConfig) error {
-	return nil
-}
-
-func (a *AzureAuth) getWorkspaceAccessToken(config *service.DBApiClientConfig) error {
-	log.Println("[DEBUG] Creating workspace token")
-	apiLifeTimeInSeconds := int32(600)
-	comment := "Secret made via SP"
-	url := "https://" + a.TokenPayload.AzureRegion + ".azuredatabricks.net/api/2.0/token/create"
-	payload := struct {
-		LifetimeSeconds int32  `json:"lifetime_seconds,omitempty"`
-		Comment         string `json:"comment,omitempty"`
-	}{
-		apiLifeTimeInSeconds,
-		comment,
-	}
-	headers := map[string]string{
-		"Content-Type": "application/x-www-form-urlencoded",
-		"X-Databricks-Azure-Workspace-Resource-Id": a.AdbWorkspaceResourceID,
-		"X-Databricks-Azure-SP-Management-Token":   a.ManagementToken,
-		"cache-control":                            "no-cache",
-		"Authorization":                            "Bearer " + a.AdbPlatformToken,
-	}
 
-	var responseMap map[string]interface{}
-	resp, err := service.PerformQuery(config, http.MethodPost, url, "2.0", headers, true, true, payload, nil)
-	if err != nil {
-		return err
-	}
-	err = json.Unmarshal(resp, &responseMap)
+	err = platformToken.EnsureFresh()
 	if err != nil {
 		return err
 	}
-	a.AdbAccessToken = responseMap["token_value"].(string)
+	a.AdbPlatformToken = platformToken.OAuthToken()
 	return nil
 }
 
 func (a *AzureAuth) initWorkspaceAndGetClient(config *service.DBApiClientConfig) (service.DBApiClient, error) {
 	var dbClient service.DBApiClient
+
+	// Get management token
 	err := a.getManagementToken(config)
 	if err != nil {
 		return dbClient, err
 	}
 
+	// Get workspace access token
 	err = a.getWorkspaceID(config)
 	if err != nil {
 		return dbClient, err
 	}
+
+	// Get platform token
 	err = a.getADBPlatformToken(config)
 	if err != nil {
 		return dbClient, err
 	}
-	err = a.getWorkspaceAccessToken(config)
-	if err != nil {
-		return dbClient, err
-	}
 
 	var newOption service.DBApiClientConfig
-	newOption.Token = a.AdbAccessToken
+	// Host for azure
 	newOption.Host = "https://" + a.TokenPayload.AzureRegion + ".azuredatabricks.net"
+	// Host for platform token
+	newOption.Token = a.AdbPlatformToken
+	// Headers to use aad tokens
+	newOption.DefaultHeaders = map[string]string{
+		"X-Databricks-Azure-Workspace-Resource-Id": a.AdbWorkspaceResourceID,
+		"X-Databricks-Azure-SP-Management-Token":   a.ManagementToken,
+	}
 	dbClient.SetConfig(&newOption)
 
-	_, err = dbClient.Clusters().ListNodeTypes()
+	// So when the workspace is initially created sometimes it fails to perform api calls so this is a simple test
+	// to verify that the workspace has been created successfully. The retry is intentional as sometimes after workspace
+	// creation the API's will not work correctly. This may also
+	err = retry(5, 1*time.Second, func() error {
+		_, err = dbClient.Clusters().ListNodeTypes()
+		return err
+	})
+
 	return dbClient, err
 }
+
+func retry(numAttempts int, sleep time.Duration, do func() error) error {
```

Comment:
This retry should help fix #33 as we will attempt to retry and above we will be retrying for about 1 minute to attempt to list the node types. This should verify that the apis are ready for requests. This is a hack, but I am internally trying to identify the root cause of why the apis are not initialized properly.